src/analysis/normed_space/inner_product.lean

```diff
@@ -0,0 +1,1553 @@
+/-
+Copyright (c) 2019 Zhouhang Zhou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Zhouhang Zhou, Sébastien Gouëzel, Frédéric Dupuis
+-/
+
+import linear_algebra.bilinear_form
+import linear_algebra.sesquilinear_form
+import analysis.special_functions.pow
+import topology.metric_space.pi_Lp
+import data.complex.is_R_or_C
+
+/-!
+# Inner Product Space
+
+This file defines inner product spaces and proves its basic properties.
+
+An inner product space is a vector space endowed with an inner product. It generalizes the notion of
+dot product in `ℝ^n` and provides the means of defining the length of a vector and the angle between
+two vectors. In particular vectors `x` and `y` are orthogonal if their inner product equals zero.
+
+## Main results
+
+- We define the class `inner_product_space` with a number of basic properties, most
+  notably the Cauchy-Schwarz inequality.
+- We show that if `f i` is an inner product space for each `i`, then so is `Π i, f i`
+- We define `euclidean_space 𝕜 n` to be `n → 𝕜` for any `fintype n`, and show that
+  this an inner product space.
+- Existence of orthogonal projection onto nonempty complete subspace:
+  Let `u` be a point in an inner product space, and let `𝕜` be a nonempty complete subspace.
+  Then there exists a unique `v` in `𝕜` that minimizes the distance `∥u - v∥` to `u`.
+  The point `v` is usually called the orthogonal projection of `u` onto `𝕜`.
+
+## Notation
+
+We locally denote the inner product by `⟪·, ·⟫`. We also provide two notation namespaces:
+`real_inner_product_space` and `complex_inner_product_space`, which introduce the notation
+`⟪·, ·⟫` for the real and the complex inner product respectively.
+
+## Implementation notes
+
+We define both the real and complex cases at the same time using the `is_R_of_C` typeclass.
+We choose the convention that inner products are conjugate linear in the first argument and linear
+in the second.
+
+## TODO
+
+- Fix the section on the existence of minimizers and orthogonal projections to make sure that it
+  also applies in the complex case.
+
+## Tags
+
+inner product space, norm
+
+## References
+*  [Clément & Martin, *The Lax-Milgram Theorem. A detailed proof to be formalized in Coq*]
+*  [Clément & Martin, *A Coq formal proof of the Lax–Milgram theorem*]
+
+The Coq code is available at the following address: <http://www.lri.fr/~sboldo/elfic/index.html>
+-/
+
+noncomputable theory
+
+open is_R_or_C real
+open_locale big_operators classical
+
+variables {F : Type*} {G : Type*}
+variables {𝕜 : Type*} [is_R_or_C 𝕜]
+local notation `𝓚` := @is_R_or_C.of_real 𝕜 _
+
+/-- Syntactic typeclass for types endowed with an inner product -/
```

Comment:
There is a difficulty here that, when you write `inner x y`, Lean can not guess the field. If you have a (complex) hermitian vector field, and you derive the real inner product space structure, then you will have two different inner spaces on it and Lean will get confused. You have a nice solution below, by introducing the notations `⟪x, y⟫_ℝ` and `⟪x, y⟫_ℂ`. I think it is worth moving these notations just after the definition and making them global. Also, introducing localized notations for `⟪x, y⟫` as you have done at the end of the file is very good, but you could move them right here. And you could also introduce a third locale `inner_product_space` where the field is unspecified, and left to Lean. Then the user could choose which locale he wants to open depending on the context. (In this file, you would want to `open inner_product_space`, for instance) (Maybe you will need to declare the notations in a section, to avoid them being open automatically in the whole file)

Reply:
I made the `⟪x, y⟫_ℝ` and `⟪x, y⟫_ℂ` declarations global and moved the rest of the notations up here, but for the `inner_product_space` localized notation, I suspect that if this is needed elsewhere, one would want to do what I did here: first declare `𝕜` to be some `is_R_or_C` field, and then declare `⟪x, y⟫` to be the inner product with this specific `𝕜` as the field. I had tried leaving it unspecified and it caused problems.